### DISCUSSION TITLE
Multi Language Projects

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#669] [Q&A] [answered]
I read https://github.com/langium/langium/discussions/642 and https://github.com/langium/langium/pull/311in the context of multi language projects (langium 0.4.0). I wonder what a good starting point would be (an example project).

My thoughts so far: **I have two DSLs**.
**(1) a DSL for requirements modeling** requirements with a name and text:

```
req1: "the thing should do this" 
req2: "the thing should do that" 
```

**(2) a DSL modeling test cases**:

```
tst1: "tst1.py"
tst2: "tst2.py"
```

now I want the testcase DSL to reference elements from the requirement DSL:

```
tst1: "tst1.py" tests req1
tst2: "tst2.py" tests req2
```

How would I achieve this in langium?
 * I want two languages (as separate as possible)
 * I want cross-language references

what impact does this have on the DSL architecture? how do services from both DSL interact? How many language servers would I have? How to create an extension or extensions for vscode...

With a short introduction I could try to create the above example as langium example with some accompanying docs:
 * I would create a single project with the two languages inside?
 * I would create a single extension?

I was confused by the comment in https://github.com/langium/langium/pull/311 "The changes can be tested with the statemachine example package.": How is this meant?

**Comment by spoenemann** [comment, marked as answer]
If the two languages are totally separate, they can be handled by different language servers (and different vscode extensions). But if you want cross-language references, they need to be processed together (in the same langium-config.json) and run in the same language server. The language server disambiguates requests based on the file extension of incoming URIs.

What you mentioned about requirements and tests sounds like a good example that we could add to `examples` in this repo. Ideally with the same structure as the other examples, but with two `.langium` files in it (every language needs a dedicated entry grammar). The example might be overly simplistic though, maybe we can make the modeling of test cases more sophisticated than just delegating to Python files?

**Reply by goto40** [reply]
I will prepare something to review.

**Reply by goto40** [reply]
~~I setup an initial working demo (based on the tag `0.4.0`). I still have questions and would like to to discuss this (in a PR). However, I have issues rebasing my branch on the current main: I get an error while generating code from my grammar using the current main:~~
```
Reading config from langium-config.json
TypeError: Cannot read properties of undefined (reading '$container')
```
~~This is possibly related to the fact that I `import` grammars into other grammars... Do you have any ideas?~~

~~My rebased branch is https://github.com/goto40/langium/tree/feature/multi-language-demo-current_main (I did not yet update my example to the newest version - because I got stuck while running langium:generate...)~~
 (solved in #672 )

**Reply by goto40** [reply]
I created https://github.com/langium/langium/issues/671.
